### PR TITLE
[JENKINS-48924, JENKINS-48995] - Fix plugin compatibility with Jenkins 2.102+

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] TestLinkBuildAction and TestLinkResult do not longer serialize build instance
- [x] Whitelist TestLink API model classes, which are being persisted in reports
- [x] Fix the Builder in ResultTestCaseSeeker class

https://issues.jenkins-ci.org/browse/JENKINS-48924
https://issues.jenkins-ci.org/browse/JENKINS-48995

@reviewbybees @jglick

### CODE REVIEW

File: src/main/java/hudson/plugins/testlink/TestLinkBuildAction.java [hunk 1]
```diff
@@ -25,32 +25,52 @@
...
 
+import jenkins.model.RunAction2;
 import org.kohsuke.stapler.StaplerProxy;
 
 /**
  * @author Bruno P. Kinoshita - http://www.kinoshita.eti.br
  * @since 1.0
  */
-public class TestLinkBuildAction implements Action, Serializable, StaplerProxy {
+public class TestLinkBuildAction implements RunAction2, Serializable, StaplerProxy {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/testlink/TestLinkBuildAction.java]: Surprisingly, #25 does not do that

**Comment by oleg-nenashev** [comment]
I currently cannot say whether it work on agents actually, the autotests do not seem to cover that

**Comment by oleg-nenashev** [comment]
PCT passes with this fix BTW

File: src/main/java/hudson/plugins/testlink/TestLinkBuildAction.java [hunk 2]
```diff
@@ -25,32 +25,52 @@
...
 
     private static final long serialVersionUID = -914904584770393909L;
 
     public static final String DISPLAY_NAME = "TestLink";
     public static final String ICON_FILE_NAME = "/plugin/testlink/icons/testlink-24.png";
     public static final String URL_NAME = "testLinkResult";
 
-    private AbstractBuild<?, ?> build;
+    private transient Run<?, ?> build;
```
Comment (jglick) [inline src/main/java/hudson/plugins/testlink/TestLinkBuildAction.java]: Good, making this `transient` should solve a host of issues.

**Comment by oleg-nenashev** [comment]
Also fixed TestNG logic, which serialization has not been covered by tests
@jglick @kinow PTAL

File: src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java [hunk 3]
```diff
@@ -58,7 +57,7 @@
 	
 	public static final String TEXT_XML_CONTENT_TYPE = "text/xml";
 
-	protected final TestNGParser parser = new TestNGParser();
+	protected transient TestNGParser _parser = null;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java]: The class could be actually made static, but it needs update of the bundled lib

File: src/test/java/hudson/plugins/testlink/result/ResultSeekerTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.For;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.kohsuke.stapler.RequestImpl;
+import org.mockito.Mock;
+
+import javax.management.Descriptor;
+
+@For(ResultSeeker.class)
+@Ignore
```
Comment (jglick) [inline src/test/java/hudson/plugins/testlink/result/ResultSeekerTest.java]: specify a reason in the annotation value

File: src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java (hunk 3 above)
Comment (jglick) [reply src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java]: Maybe should be `volatile`.

File: src/test/java/hudson/plugins/testlink/result/ResultSeekerTest.java (hunk 4 above)
Comment (oleg-nenashev) [reply src/test/java/hudson/plugins/testlink/result/ResultSeekerTest.java]: Oops, I pushed an unfinished test.
It does not hurt in the current state, but I will remove the file

File: src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/plugins/testlink/result/AbstractTestNGResultSeeker.java]: Does not matter much. In the worst case there will be two instances of a same class which actually has no dynamic context at all. The class methods should be reworked to static, I just decided not to do breaking changes in this PR

**Comment by oleg-nenashev** [comment]
@kinow I noticed the plugin is up for adoption. Do you have time to review/release it or should we proceed on our own?

**Comment by kinow** [comment]
@oleg-nenashev either way works for me. I stopped keeping up to date with Jenkins plugin/core API when the workflow-plugin was still in the works, and haven't had time to catch up with the changes.

Looks to me like most of the changes are for the JEP and Jenkins/plug-in API. No changes in the way the plug-in interacts with TestLink or test reports. So everything should work the same for users (just much better, and in the latest versions of Jenkins :-) ).

So +1, LGTM, and thanks a lot!

Happy to merge it, or feel free to merge it. Do we need #25 merged as well?

If you'd like to take over the plug-in, I'd be OK with that too :) tho I suspect you are probably busy with other things and just needs to fix this issue and have a new release out?

Ta
B

**Comment by oleg-nenashev** [comment]
Sorry, responded from wrong email

**Comment by kinow** [comment]
Sure, works for me @oleg-nenashev.

What about the #25? Is that required as well for a release?

**Comment by kinow** [comment]
Will cut a release as is (and plus #25 if @oleg-nenashev says it needs to be included for this PR) this weekend (this is an extended weekend, with Monday being Auckland anniversary, so will either release Friday night, or Monday evening NZ time).

**Comment by oleg-nenashev** [comment]
I re ommend to release without #25, it needs more reviews imo

On Jan 25, 2018 01:09, "Bruno P. Kinoshita" <notifications@github.com>
wrote:

Will cut a release as is (and plus #25
<https://github.com/jenkinsci/testlink-plugin/pull/25> if @oleg-nenashev
<https://github.com/oleg-nenashev> says it needs to be included for this
PR) this weekend (this is an extended weekend, with Monday being Auckland
anniversary, so will either release Friday night, or Monday evening NZ
time).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jenkinsci/testlink-plugin/pull/29#issuecomment-360317133>,
or mute the thread

**Comment by kinow** [comment]
Thanks for the quick reply. Had some spare time tonight, enough to cut a new release. Just released it, 3.13 should make the update centre in the next hours I believe.

Let me know if you need anything else, and thanks for working on it 👍 
B

**Comment by oleg-nenashev** [comment]
@kinow Thanks a lot, I will update the JEP-200 status page. If you have some time, review/release of https://github.com/jenkinsci/tap-plugin/pull/20 would be great

**Comment by kinow** [comment]
Oh! Was going to forget that one and probably look into that only next week. Merged, preparing a release now :-)

**Pull request merged**